packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2894,7 +2920,15 @@ export class ContainerRuntime
 				submitOpDuration: trace.trace().duration,
 			} as const;
 
-			this.summarizerNode.completeSummary(handle);
+			try {
+				// If validateSummaryBeforeUpload is false, the summary should be validated in this step.
+				this.summarizerNode.completeSummary(
+					handle,
+					!this.validateSummaryBeforeUpload /* validate */,
+				);
+			} catch (error) {
+				return { stage: "upload", ...uploadData, error };
```

Comment:
Can we add a different stage? This is like post upload validation, if not, no big deal

Reply:
Or even if it's not called a different stage, could we just call `validate` here, instead of merging it into `completeSummary`?  On the other hand, you already wrote the code this way so it's ok.

Reply:
The reason for not calling `validate` here is that it does a DFS of the entire tree. So, we will end up doing this DFS twice - once for `validate` and once for `complete` which is unncessary.

Reply:
As for the stage, it really should be `submit` with an error indicating summarization failed but it's just too much plumbing and not worth it really. Same goes for adding a new stage. The code to do this is just too complicated - I will have to add new interfaces, new success and failure types and those are all breaking changes.